crypto/evp/p_lib.c

```diff
@@ -80,66 +81,141 @@ int EVP_PKEY_save_parameters(EVP_PKEY *pkey, int mode)
+            && (EVP_KEYMGMT_number(to->pkeys[0].keymgmt)
+                != EVP_KEYMGMT_number(from->pkeys[0].keymgmt))) {
+            EVPerr(EVP_F_EVP_PKEY_COPY_PARAMETERS, EVP_R_DIFFERENT_KEY_TYPES);
             return 0;
-    } else if (to->type != from->type) {
-        EVPerr(EVP_F_EVP_PKEY_COPY_PARAMETERS, EVP_R_DIFFERENT_KEY_TYPES);
-        goto err;
+        }
+        is_provided = 1;
+    } else {
+        if (to->type == EVP_PKEY_NONE) {
+            if (EVP_PKEY_set_type(to, from->type) == 0)
+                return 0;
+        } else if (to->type != from->type) {
+            EVPerr(EVP_F_EVP_PKEY_COPY_PARAMETERS, EVP_R_DIFFERENT_KEY_TYPES);
+            return 0;
+        }
     }
 
     if (EVP_PKEY_missing_parameters(from)) {
         EVPerr(EVP_F_EVP_PKEY_COPY_PARAMETERS, EVP_R_MISSING_PARAMETERS);
-        goto err;
+        return 0;
     }
 
     if (!EVP_PKEY_missing_parameters(to)) {
+        /*
+         * TODO(3.0) EVP_PKEY_cmp_parameters() may be inefficient for
+         * comparison of provided domain parameters, can we make it better?
+         */
         if (EVP_PKEY_cmp_parameters(to, from) == 1)
             return 1;
         EVPerr(EVP_F_EVP_PKEY_COPY_PARAMETERS, EVP_R_DIFFERENT_PARAMETERS);
         return 0;
     }
 
-    if (from->ameth && from->ameth->param_copy)
+    /*
+     * If |from| is provided, we ONLY copy the provided bits, and therefore
+     * clear the legacy bits in |to|.
+     */
+    if (is_provided) {
+        evp_pkey_free_legacy(to);
+        return evp_keymgmt_copy(to, from, 1);
+    }
+
+    /* Legacy support */
+    if (from->ameth != NULL && from->ameth->param_copy != NULL)
         return from->ameth->param_copy(to, from);
- err:
+
     return 0;
 }
 
 int EVP_PKEY_missing_parameters(const EVP_PKEY *pkey)
 {
-    if (pkey != NULL && pkey->ameth && pkey->ameth->param_missing)
-        return pkey->ameth->param_missing(pkey);
+    if (pkey != NULL) {
+        if (pkey->ameth == NULL)
+            return !evp_keymgmt_is(pkey, 1);
+        if (pkey->ameth->param_missing != NULL)
+            return pkey->ameth->param_missing(pkey);
+    }
     return 0;
 }
 
+static int cmp_provided(const EVP_PKEY *a, const EVP_PKEY *b, int domainparams)
+{
+    /*
+     * TODO: when legacy is gone, we can remove this function and replace
+     * calls to it with calls to evp_keymgmt_cmp().
+     */
+
+    /* If none of the keys have a non-NULL keymgmt, it's pointless to try */
+    if (a->pkeys[0].keymgmt == NULL && b->pkeys[0].keymgmt == NULL)
+        return -2;
+
+    /*
+     * Ensure on the fly that they both have a keymgmt if at least one of
+     * them does.
+     * To do so, we must break constness to be able to cache keymgmt data,
+     * but since we know that EVP_PKEYs are dynamically allocated, it's ok.
+     */
+    if (!evp_pkey_make_provided((EVP_PKEY *)a, NULL, NULL, NULL, domainparams)
+        || !evp_pkey_make_provided((EVP_PKEY *)b, NULL, NULL, NULL,
```

Comment:
Again, I'm concerned about the use of the default libctx here. What if the default libctx doesn't have a provider for this type of key loaded?